test-utils/src/dev/dendrite.rs

```diff
@@ -0,0 +1,148 @@
+            .join("..")
+            .join("out")
+            .join("dendrite-stub")
+            .join("bf_sde");
+
+        let args = vec![
+            "run".to_string(),
+            "--listen-addresses".to_string(),
+            address_one,
+        ];
+
+        let child = tokio::process::Command::new("dpd")
+            .args(&args)
+            .stdin(Stdio::null())
+            .stdout(Stdio::from(redirect_file(
+                temp_dir.path(),
+                "dendrite_stdout",
+            )?))
+            .stderr(Stdio::from(redirect_file(
+                temp_dir.path(),
+                "dendrite_stderr",
+            )?))
+            .env("P4_DIR", p4_dir)
+            .spawn()
+            .with_context(|| {
+                format!("failed to spawn `dpd` (with args: {:?})", &args)
+            })?;
+
+        let child = Some(child);
+
+        if port == 0 {
+            port = discover_port(
+                temp_dir.path().join("dendrite_stdout").display().to_string(),
+            )
+            .await?;
+        }
+
+        Ok(Self { port, args, child, data_dir: Some(temp_dir) })
+    }
+
+    pub async fn cleanup(&mut self) -> Result<(), anyhow::Error> {
+        if let Some(mut child) = self.child.take() {
+            child.start_kill().context("Sending SIGKILL to child")?;
+            child.wait().await.context("waiting for child")?;
+        }
+        if let Some(dir) = self.data_dir.take() {
+            dir.close().context("Cleaning up temporary directory")?;
+        }
+        Ok(())
+    }
+}
+
+impl Drop for DendriteInstance {
+    fn drop(&mut self) {
+        if self.child.is_some() || self.data_dir.is_some() {
+            eprintln!(
+                "WARN: dropped DendriteInstance without cleaning it up first \
+                (there may still be a child process running and a \
+                temporary directory leaked)"
+            );
+            if let Some(child) = self.child.as_mut() {
+                let _ = child.start_kill();
+            }
+            if let Some(dir) = self.data_dir.take() {
+                let path = dir.into_path();
+                eprintln!(
+                    "WARN: dendrite temporary directory leaked: {}",
+                    path.display()
+                );
+            }
+        }
+    }
+}
+
+fn redirect_file(
+    temp_dir_path: &Path,
+    label: &str,
+) -> Result<std::fs::File, anyhow::Error> {
+    let out_path = temp_dir_path.join(label);
+    std::fs::OpenOptions::new()
+        .write(true)
+        .create_new(true)
+        .open(&out_path)
+        .with_context(|| format!("open \"{}\"", out_path.display()))
+}
+
+async fn discover_port(logfile: String) -> Result<u16, anyhow::Error> {
+    let timeout = Instant::now() + Duration::new(5, 0);
+    tokio::time::timeout_at(timeout, find_dendrite_port_in_log(logfile))
+        .await
+        .context("time out while discovering dendrite port number")?
+}
+
+async fn find_dendrite_port_in_log(
```

Comment:
This will easily break if dpd log formatting changes. Can we determine the port in the test harness and pass it to dpd as an argument instead?

Reply:
It might be difficult to do so reliably. We are using the same pattern for generating port numbers for `cockroach` and `clickhouse` (using port `0` to have the OS dynamically determine the port number), and I don't think there's any accounting going on to keep track of port numbers that have already been used. It might be easier to steal a page from how `cockroach` works and have `dpd` generate a file containing the connection information in a consistent format on startup.

Reply:
I use the `portpicker` crate [in falcon](https://github.com/oxidecomputer/falcon/blob/main/lib/src/lib.rs#L475-L478) to do this for launching propolis daemons. This uses information from the OS to pick the port, returning the selected port directly to the driving code to deliver to the daemon being launched. Much less error-prone the log spelunking.

Reply:
The `portpicker` crate looks like it has an pretty inherent TOCTOU race in it. That's why we've generally looked back at the server after it's started to see which port the OS actually allocated to it. Especially since we're launching potentially many of these during a test run, we'd at least want to try to detect collisions and restart the server in a loop.

Reply:
:+1: on avoiding `portpicker`; its TOCTOU race was the source of the test flake in #1200.

Reply:
Sorry for the bad suggestion. But I do think we should have a follow-up issue to not depend on regexing logs to determine ports.

Reply:
No worries, and agreed, regex'ing is kinda brittle. 

> It might be easier to steal a page from how cockroach works and have dpd generate a file containing the connection information in a consistent format on startup.

Do you think this is a good approach? This resolves the TOCTOU and log-scraping-brittleness concerns.

Reply:
Yeah, I think that's fine.